doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+
+The **compute layer** (`dataflow`, plus potentially others) consumes streams of updates to collections, and computes and then incrementally maintains the results of views over these data.
+A view is ideally "deterministic" in the sense that it acts as a function on the input update streams, and produces a well-defined output on definite input (we can allow "non-deterministic" views but like "indefinite streams" they taint their results).
+One correctness goal of the `dataflow` module is to convert definite inputs into definite outputs, which can be returned to the storage layer or to interactive user sessions.
+Views and their dependencies form the "concurrency control" of Materialize, in that outputs are reported as definite only when, and ideally as soon as, their inputs are definite.
+
+The **coordination layer** (`coord`, plus potentially others) interfaces with user sessions and pilots the storage and compute layers.
+It maintains mappings from user names to storage and compute objects (source, views, indexes), and dereferences them in ways that provide a consistent experience for users.
+It also chooses timestamps for reads (peeks) and writes (table inserts) to provide the lived experience of continually moving forward through time.
+It pilots the storage and compute layers and is responsible for ensuring they communicate clearly and completely with each other.
+One correctness goal of the `coord` module is to provide "strict sequential consistency" to the interactive users, through the careful use of the storage and compute layers.
+Another correctness goal of the `coord` module is to ensure that the storage and compute correctly and completely interoperate (e.g. that the compute layer is invoked only on definite inputs, and fully communicates its definite outputs back to storage if appropriate).
+
+### Approaches
+
+The plan of record is that source data are made definite by the storage layer, and results are definite when observed through the compute layer (applied to definite inputs).
+Definite data are both durable, and ordered in time (strictly sequentially consistent), and reflect the stated "order" the system presents as having undergone.
+
+The coordination layer has several constraints on its behavior to present as strictly sequentially consistent.
+The coordination layer should only report definite data, or indefinite data that are clearly presented as such.
+The coordination layer's catalog should advance in a similar timeline as the data it references, to avoid apparent mis-ordering of DDL and DML statements.
+The coordination layer may need to wait on data becoming definite when returning from e.g. INSERT statements where the return communicates durability.
+
+Data that are provided as update streams, not through interactive sessions, are treated as occurring at the time assigned to them.
+They are not subject to the real-time requirements of strict sequential consistency in the same way that foreign tables are not.
```

Comment:
It is not clear to me what this means - can you give an example permitted anomaly?

Reply:
Right, good point. This came up in many discussions with @mjibson but I forget that it is not shared context. Guarantees like "linearizability" speak about partial orders of events, defined by their `[start, end]` intervals in real time. If an event has ended before a next one starts, the next one must reflect the first.

This is a bit weird with explicitly timestamped data, which announce when they should be seen as happening, even though the actual e.g. read from Kafka could have been a long time before or after that time. There is a legit Q about "well, for linearizability of these actions, do we need to track the moment that data landed in Kafka and hold back all of our results?". That would be super helpful, as it means Kafka+MZ provides some serious guarantees. But it is also much harder to do, especially if we don't have explicit hooks into these adjacent bits of tech to either hold them or us up until a thing is done.

Alternatively, explicitly timestamped things say "independent of when I actually happen, you should pretend that I happened at this time". I think this is justified, because these are notes that essentially say "I could literally wait until this time, and then "happen" but that is a waste of all of our time and energy; instead, I'll tell you ahead of time when I will occur." I think of it as a compressed, data-driven representation of what is otherwise a behavioral constrant.

So an allowed anomaly would be
```
FRANK @ 1: Reads at 1, get FOO.
KAFKA @ 2: Writes data with timestamp 7.
FRANK @ 3: Reads at 3, get FOO (doesn't reflect above write).
```
There are probably other anomalies, in particular wrt sources that hold back time rather than speaking about the future. I have a harder time thinking about "allowed anomalies" and am better at thinking about "correct, prescribed outputs", though.